app/ocr.py

```diff
@@ -384,128 +383,175 @@ def get_city_state(self):
+                if(statute_shape_match[0:4] == 'ddd(' or statute_shape_match[0:4] == 'ddd.' or statute_shape_match[0:5] == 'dddd.' or statute_shape_match[0:5] == 'dddd(' or statute_shape_match[0:6] == 'ddddd.' or statute_shape_match[0:6] == 'ddddd('):                
+                    # Close any open parentheses & append if not already present
+                    has_open_paren = False
+                    temp_token3 = str(token)
+                    for i in range(0, len(temp_token3)):
+                        if(temp_token3[i] == '('):
+                            has_open_paren = True
+                        if(temp_token3[i] == ')'):
+                            has_open_paren = False
+                    if(has_open_paren == True):                    
+                        temp_token3 = temp_token3 + ')'
+                    if temp_token3 not in statutes_list:
+                        statutes_list.append(temp_token3)
+        statutes_list = sorted(statutes_list)
+        
+        # Creates a dictionary with the key being the type of statute, and value being the listed statutes determined by the first 4 numbers in a statute.
+        return_dict = {}
+        CFR = []
+        INA = []
+        other = []
+        USC = []
+        for j in range(0, len(statutes_list)):
+            if(statutes_list[j][0:4].isnumeric() == True):
+                if(int(statutes_list[j][0:4]) >= 1000 and int(statutes_list[j][0:4]) <= 1003):
+                    CFR.append(statutes_list[j])
+                    continue
+                elif(int(statutes_list[j][0:4]) >= 1100 and int(statutes_list[j][0:4]) <= 1999):
+                    USC.append(statutes_list[j])
+                    continue
+                else:
+                    other.append(statutes_list[j])
+                    continue        
+            elif(statutes_list[j][0:3].isnumeric() == True):
+                if(int(statutes_list[j][0:3]) >= 100 and int(statutes_list[j][0:3]) <= 199):
+                    USC.append(statutes_list[j])
+                    continue
+                elif(int(statutes_list[j][0:3]) >= 200 and int(statutes_list[j][0:3]) <= 399):
+                    INA.append(statutes_list[j])
+                    continue
+                else:
+                    other.append(statutes_list[j])
+                    continue
+            else:
+                other.append(statutes_list[j])
+        #TODO: Remove overlap between corresponding INA and USC statutes, recreate this table- https://www.uscis.gov/laws-and-policy/legislation/immigration-and-nationality-act
+        #TODO: Make a dictionary that has the laws for a corresponding statute, map those to a certain level of granularity that can be displayed to the end user
+        return_dict["CFR"] = CFR
+        return_dict["INA"] = INA
+        return_dict["Other"] = other
+        return_dict["USC"] = USC
+        return return_dict

```

Comment:
Based on the 'TODO' included I think this function will need more work, but it is a great start to retrieving the referenced statutes in the case.

Reply:
Once more work is put into refining this, I think it can be a stronger identifer of both protected classes and why the person is seeking asylum in the first place. Plus with using a custom scraper it's less accurate on individual words, so we would either need to drop misreads or flag them for manual review.